### DISCUSSION TITLE
Тестовый сайт RuCore2 для проверки генерации IG

### DISCUSSION THREAD

**Discussion by eugenekogan-spb** [fhir-ru/core#230] [General]
Итоговая карточка задачи по состоянию на 13 октября 2023
_____________________________________________________________
На [встрече 25 июля](https://fhir.ru/meetings.2023-07-25) и в последующих обсуждениях решили подготовить мини-комплект артефактов FHIR для отладки компиляции IG, разместить их на страницах сайта, сделать комплекты руководств. Будем тестировать:
- решения по размещению текстов на страницах сайта
- решения по именованию артефактов и прочим соглашениям
- сам механизм редактирования сайта и публикации руководств

1) Принимаем, что в комплект следует включить артефакты:
- CorePatient: Patient
-- identifier, slice по system включаем только Государственное свидетельство о рождении
-- name
-- address, тип данных CoreAddress
-- пример экземпляра CorePatient

- CoreServiceRequest: ServiceRequest
-- Нет своих ограничений

- CoreAddress: Address 
-- расширения для ФИАС

- LabServiceRequest: CoreServiceRequest
-- subject -> CorePatient 
-- пример экземпляра LabServiceRequest с заполненной ссылкой на пример CorePatient

- NamingSystem для системы выдачи государственных свидетельств о рождении
- NamingSystem, CodeSystem, ValueSet для Справочника "Документы, удостоверяющие личность"

2) Артефакты размещаем на сайте https://fhir-ru.zendoc.me, на следующих страницах:
- Руководство RuCore
   -- Профили ресурсов
      --- CorePatient (содержит fsh-тексты профиля, экземляра ресурса)
   -- Профили типов данных
      --- CoreAddress (содержит fsh-текст профиля)
   -- Расширения
      --- CoreAddressFIASHouse (содержит fsh-текст расширения)
   -- Наборы значений
      --- CoreVsIdentityDocument (содержит fsh-текст набора значений, ссылку на страницу системы кодирования) 
   -- Системы кодирования
      --- CoreVsIdentityDocument  (содержит fsh-тексты системы именования и системы кодирования, ссылку на внешний справочник, ссылку на набор значений)
   -- Системы именования
      --- FIAS (содержит fsh-текст системы именования)
- Руководство СЭМД ПЛИ
   -- Профили ресурсов
      --- LabServiceRequest (содержит fsh-тексты профиля, экземляра ресурса)

3) Система публикации руководств (пока включает ручную работу Тимура) создает руководства:
- RuCore https://rucore.fhir.ru
- RuLab https://rulab.fhir.ru

4) Проверяем следующие утверждения:
- [ ] CorePatient, NamingSystem FIAS, ValueSet/CodeSystem/NamingSystem CoreVsIdentityDocument включились в руководство RuCore
- [ ] LabServiceRequest включен в руководство RuLab
- [ ] В руководстве RuLab работает ссылка LabServiceRequest.patient на страницу CorePatient руководства RuCore
- [ ] Работают ссылки между страницами внутри руководств
- [ ] Работают ссылки между страницами сайта, для удобства навигации при редактировании сайта

**Comment by eugenekogan-spb** [comment]
На [встрече 1 августа](https://fhir.ru/meetings.2023-08-01) решили, для полноты тестирования сайта: 

- сократить объем описания пациента, чтобы не задействовать ссылки, которые могут не работать
- добавить недостающие типы артефактов - NamingSystem, Extension
- добавить примеры ресурсов

**Reply by Vrodvi** [reply]
- сократить объем описания пациента, чтобы не задействовать ссылки, которые могут не работать
  - можно смело сокращать фрагменты со следующими идентификаторами:
    - MRN 0..1 - номер мед.карты у каждой информационной системы будет свой, в своём пространстве идентификаторов, единого пространства номеров карт на все системы быть не может, следовательно такой срез не возможен, в этом месте будет бесконечное множество срезов номеров карт для каждой из систем
      - аналогично для внутреннего идентификатора пациента PatientInternalIdentifier
      - с этой проблемой в СЭМДе боролись выдачей рекомендаций по формату уникальных пространств идентификаторов для разных сущностей. Этот вопрос явно тянет на отдельное обсуждение.
    - InsuranceOMS - полиса есть старого образца, нового образца, есть временные. Что тут имеется ввиду не ясно. А единого пространства на идентификаторы полисов всех типов быть не может.
   


- добавить недостающие типы артефактов - NamingSystem, Extension
  - NamingSystem нужно добавить для всех справочников, использованных в профилях, а также для всех пространств идентифакаторов, использованных в атрибутах Identifier
  - Extension рекомендуется добавить тот, который добавляет атрибуты с кодами ФИАСа в адресе. Заодно нужно определить профиль на локальный тип данных для адреса RU-core, и в нём уже использовать расширение. Таким образом мы покроем максимальное количество профилируемых типов ресурсов.
  
- добавить примеры ресурсов
  - обязательно нужен максимально полный пример, в котором будут заполнены все обозначенные в профиле поля. Желательно, чтобы при этом пример представлял из себя осмысленное наполнение из реального мира. Если реальный пример не помещается в полную структуру примера, то нужно делать два примера - полный технически заполненный и второй реальный пример, заполненный на сколько это возможно. Также желательно иметь ещё и минимально заполненный пример, соответствующий профилю, в котором будут заполнены только обязательные поля.

**Comment by Vrodvi** [comment]
По поводу вот этих разделов:

Наборы значений
-- Виды полиса обязательного медицинского страхования (текст fsh)
Системы кодирования
-- Виды полиса обязательного медицинского страхования (текст fsh)

Согласен, что должны быть именно такие разделы, но вот что в них располагаться должно и как должны работать ссылки между ними - это просьба обсудить.

Особенность описания справочников (Систем кодирования) в том, что справочник можно описать как запись в реестре (через ресурс https://www.hl7.org/fhir/namingsystem.html) и ещё дополнить эту запись в реестре описанием самого справочника (через ресурс https://www.hl7.org/fhir/codesystem.html).

Если справочник внешний - т.е. такой который создан сторонней организацией и существует вне руководства по реализации - то его следует описывать только как реестровую запись (через ресурс https://www.hl7.org/fhir/namingsystem.html).

А внутренние справочники, определённые в самом руководстве, должны быть описаны двумя ресурсами (https://www.hl7.org/fhir/namingsystem.html и https://www.hl7.org/fhir/namingsystem.html).

Проблема возникает когда нужно сослаться из набора значений на сам справочник - куда должна вести ссылка - на реестровую запись или на описание справочника? При том что описания справочника (для внешних справочников) в виде ресурса CodeSystem может не быть.
Ресурс ValueSet, к примеру, всегда ссылается на CodeSystem, даже если такой ресурс на определён.
Вот пример из официальной спецификации FHIR, который содержит путающую навигацию, которую хотелось бы избежать в руководствах Ru-Core.
В разделе, посвящённом терминологиям (https://www.hl7.org/fhir/terminologies-systems.html) видим вкладку "CodeSystem", в ней - внешние справочники и среди них - SNOMED:
![image](https://github.com/fhir-ru/core/assets/92928137/428c233c-b780-4886-8658-c3fdfefbc707)
В данном случае строка в таблице - это не ресурс CodeSystem, а NamingSystem, и все ссылки ведёт сразу на сайт SNOMED.
Далее, если посмотреть набор значений, созданный на основе SNOMED (к примеру, https://www.hl7.org/fhir/valueset-condition-code.html), то там тоже ссылки сразу ведут на сайт SNOMED:
![image](https://github.com/fhir-ru/core/assets/92928137/e8f9f576-a61c-4974-802f-947bb919abbf)

Получается, что в спецификации FHIR нельзя попасть на некую выделенную страницу, посвящённую справочнику SNOMED, и через неё выйти на наборы значений этого справочника (при этом есть отдельный сайт с терминологиями, где snomed, внезапно уже описан как CodeSytem - https://terminology.hl7.org/CodeSystem-v3-snomed-CT.html).

На ZenDoc сейчас тоже есть эта неочевидная навигация, где из набора значений не возможно попасть на страницу с описанием справочника, а сразу дана ссылка на исходных источник справочника:
![image](https://github.com/fhir-ru/core/assets/92928137/c59799d7-7e8e-4e97-8481-bee2f7d3a0e6)

---

Моё мнение:
1) на странице с перечнем справочников ("Системы кодирования") должен быть перечень реестровых записей с возможностью перейти на страницу справочника.
Находясь в реестре справочников должна быть возможность скачать соответствующий NamingSystem каждой реестровой записи.
При переходе по реестровой записи мы должны попадать на страницу описания справочника, которая должна состоять из двух частей  - описания данных реестровой записи (т.е. NamingSystem) и описания данных самого справочника (если такое есть - т.е. CodeSystem).
Также со страницы справочника должна быть возможность перейти на все страницы наборов значений, использующих этот справочник.

2) на странице набора значений должны иметься ссылки на страницы всех справочников, из которых взяты значения. Это должны быть ссылки на страницы справочников внутри руководства, а не на страницы исходных источников справочников (т.е., к примеру, не на сайт ФНСИ).
Т.е. даже если в руководстве определён набор значений из справочника, которого нет в самом руководстве в виде ресурса CodeSystem, ссылка должна вести не на источник справочника, а на страницу с описанием этого справочника, где будут только реестровые сведения (и уже там будет ссылка на исходный источник).

---

Решение заводить или не заводить CodeSystem для всех без исключения использованных в руководстве справочников (чисто для унификации) - на усмотрение сообщества.

---

Ожидаемая схема страниц и ссылок: 

[страница 1] Наборы значений
| Набор значений | Ссылка |
| Виды полиса обязательного медицинского страхования | ссылка на [страница 2] |

[страница 2] Наборы значений -> Виды полиса обязательного медицинского страхования (текст fsh)
описание набора значений и ссылка на [страница 4]

[страница 3] Системы кодирования
| Система кодирования | Ссылка |
| Виды полиса обязательного медицинского страхования | ссылка на [страница 4] |

[страница 4] Системы кодирования -> Виды полиса обязательного медицинского страхования (текст fsh)
раздел 1 - общее введения к системе кодирования (человекочитаемая часть описания)
раздел 2 - описание реестровой части системы кодирования (сведения из NamingSystem)
раздел 3 - описание самой системы кодирования (сведения из CodeSystem)
раздел 4 - ссылки на наборы значений (ссылка на [страница 2])

**Comment by eugenekogan-spb** [comment]
Полностью согласен со схемой, описанной Виталием. Добавляя к этому задачу по описанию пространств имен для идентификаторов, не являющихся справочниками, получаем следующую картину улов и страниц сайта.

Узел «Системы идентификации»
Здесь полный реестр используемых в руководстве пространств имен идентификаторов, например паспорт РФ.
Каждая страничка в составе узла:
•	описывает одно пространство идентификаторов с помощью ресурса Naming System

Узел «Системы кодирования (Справочники)»
Здесь полный реестр используемых в руководстве справочников, как определенных в составе нашего руководства, так и внешних.
Каждая страничка в составе узла:
•	описывает один справочник с помощью ресурса NamingSystem и, при наличии, ресурса CodeSystem
•	содержит информацию о внешнем источнике данных справочника, при наличии, в том числе ссылку
•	содержит список наборов значений, в которых использован этот справочник

Узел «Наборы значений»
Здесь полный реестр используемых в руководстве наборов значений.
Каждая страничка в составе узла:
•	описывает один набор значений с помощью ресурса ValueSet 
•	содержит ссылки на справочники, из которых взяты значения для данного набора

После написания примеров страниц можно будет утвердить для страницы каждого типа состав разделов и автоматически валидировать по "схеме страницы".

**Comment by Vrodvi** [comment]
@eugenekogan-spb 

Вот, Евгений, тут всё обсуждено и учтено - что должно быть в минимальном наборе, подобраны все виды артефактов, учтём момент с двумя взаимосвязанными руководствами. 

Оно всё хоть и размазано по описанию карточки и комментариям, но все нужны фрагменты описания в этой дискуссии имеются. 
Ранее обсуждалось, что в случае обсуждения важных моментов в комментарии к дискуссии, автор дискуссии периодически собирает все эти сведения и вносит их в основное описание карточки дискуссии. 
Видимо пришло время собрать все обсуждённые момент, зафиксировать их в карточке и возобновить работы по этому направлению с учётом собранных сведений.

**Reply by eugenekogan-spb** [reply]
@Vrodvi Виталий, у нас есть варианты.

Вариант 1: настаиваем на принятом ранее решении. Я собираю требования к тестовому сайту с учетом сделанных в этой дискуссии комментариев, ждем реализации core-address на fsh и отдаем Тимуру на сборку сайта.
Вариант 2: учитываем работу, сделанную за прошедшие два месяца, и в качестве тестового рассматриваем собранный Тимуром пример на основе файлов Condition.fsh и Encounter.fsh: сайт https://fhir-ru.zendoc.me и опубликованные на его основе сайты руководств https://rucore.fhir.ru и https://rulab.fhir.ru

В варианте 2 я вижу следующие плюсы:
- сайт есть, его можно анализировать прямо сейчас
- присутствуют профили, расширения, системы кодирования, наборы значений, примеры
- присутствует профиль (LabServiceRequest), относящийся к другому руководству
- файлы CoreCondition и CoreEncounter, подготовлены квалифицированным участником сообщества, поэтому вероятность ошибок в них невелика

При этом, есть и минусы:
- на сайт выложены четыре профиля RuCore, а не один
- отсутствуют системы именования

**Reply by Vrodvi** [reply]
Вариант 1 - это чёткий набор артефактов, которые можно проверить и для которых известно ожидаемая структура и наполнение.

--- 

Вариант 2 - это кто-то что-то сделал, вроде бы можно посмотреть, но по факту много чего не работает:
- примеров на профили не подготовили, 
- ссылки между профилями разных руководств не работают (из лаб. профиля перейти по ссылке на профиль корневого пациента невозможно), 
- множество лишних элементов структуры, которые только загромождают чтение и понимание (массив идентификаторов у пациента, к примеру, - выше есть рекомендация сократить только до одного идентификатора), 
- корневые профили не используют другие корневые профили - к примеру, из пациента ссылка на организацию указана не на корневой профиль организации,
- руководства вроде бы как разведены, но нет описания как именно этого удалось достичь и следовательно непонятно как авторам профилей оформлять подобного рода разведение на разные руководства
- и другие

Что из этого должно работать и как именно - никому не известно, что делает адекватную проверку невозможной.

**Reply by eugenekogan-spb** [reply]
@Vrodvi Виталий, я собрал карточку, проверьте пожалуйста. Поскольку по поводу идентификатора "Полис ОМС" высказано замечание о том, что единой системы выдачи номеров не существует, я оставил другой идентификатор - Паспорт РФ.
В качестве образцовой системы кодирования в этом случае, видимо, может быть предложен ФИАС.
С другой стороны, его использование противоречит ранее принятому решению: расширение CoreAddressFiasHouse имеет тип string и не ссылается на какой-либо справочник - https://fhir.ru/core.extensions.core-ex-address-fias-house.

**Reply by Vrodvi** [reply]
Всё верно - ФИАС не система кодирования, это система идентификации адресных объектов, возможно более правильно было бы использовать тип данных identifier, но тут делали по аналогии с СЭМД, так что это тема отдельного обсуждения как лучше сделать.

Поэтому в качестве идентификатора пациента предлагаю взять identifier:BirthCertificate "Номер государственного свидетельства о рождении", т.к. там используется система кодирования "Документы, удостоверяющие личность" (https://nsi.rosminzdrav.ru/#!/refbook/1.2.643.5.1.13.13.99.2.48).

**Reply by eugenekogan-spb** [reply]
Исправил карточку, проверьте пожалуйста.

**Reply by Vrodvi** [reply]
Всё отлично.

Есть небольшие замечания:

Замечания по оформлению карточки:
- предлагаю из описания карточки убрать старую текстовку (всё равно она нечитаема, т.к. там просто текст FSH-файла), тем более что github ведёт историю изменений карточки и при желании старое наполнение можно посмотреть в истории

Замечания по набору ресурсов:
- для лабораторного ресурса базовый профиль должен быть Core, а не fhir-овский (считаю, что все последующие руководства должны иметь в основе только корневые профили, а не смесь из корневых и исходных fhir-овских), поэтому необходимо ещё один профиль CoreServiceRequest (без каких либо ограничений, пустой, просто чтобы можно было на него сослаться)



Для фиксации прогресса по сформированному минимальному набору артефактов и контроля вопросов, связанных с их описанием завёл табличку: 
https://docs.google.com/spreadsheets/d/196LOUi5C4c25xIAb8f_keI6I2GCyXCZw35W7pu7CIag/edit?usp=sharing

**Reply by eugenekogan-spb** [reply]
Правки внес. Из карточки старый текст уберу, но чуть позже - сейчас пусть будет под рукой для создания нового.
Если LabServiceRequest ссылается на CoreServiceRequest, получается не совсем правдоподобная картина, потому что единственное включенное в него решение по профилированию (subject: CorePatient) должно быть включено в CoreServiceRequest. Для целей тестирования это, конечно, не важно.

**Reply by Vrodvi** [reply]
Да, согласен, значит LabServiceRequest сокращается до пустого профиля, просто основывающегося на CoreServiceRequest (для того чтобы можно было протестировать связь между двуми руководствами этого будет достаточно).

**Comment by eugenekogan-spb** [comment]
> Если справочник внешний - т.е. такой который создан сторонней организацией и существует вне руководства по реализации - то его следует описывать только как реестровую запись (через ресурс https://www.hl7.org/fhir/namingsystem.html).

@Vrodvi Виталий, непонятно как в этой ситуации указать, на какую систему кодирования опирается набор значений. Нужен ее идентификатор, а если система кодирования не входит в наш комплект, у нее идентификатора нет.

**Reply by Vrodvi** [reply]
[namingsystem.html](https://www.hl7.org/fhir/namingsystem.html) позволяет указать целый массив идентификаторов и даже указать какой из них является предпочтительным.
